File: gym/blade/Relationships.py
from typing import Dict, List, Optional


class Relationships:
    def __init__(
        self,
        hostiles: Optional[Dict[str, List[str]]] = None,
        allies: Optional[Dict[str, List[str]]] = None,
    ):
        self.hostiles: Dict[str, List[str]] = hostiles if hostiles is not None else {}
        self.allies: Dict[str, List[str]] = allies if allies is not None else {}

    def add_hostile(self, side_id: str, hostile_id: str):
        if side_id not in self.hostiles:
            self.hostiles[side_id] = []
        if hostile_id not in self.hostiles[side_id]:
            self.hostiles[side_id].append(hostile_id)
        self.remove_ally(side_id, hostile_id)

    def remove_hostile(self, side_id: str, hostile_id: str):
        if side_id in self.hostiles:
            self.hostiles[side_id] = [
                id for id in self.hostiles[side_id] if id != hostile_id
            ]

    def add_ally(self, side_id: str, ally_id: str):
        if side_id not in self.allies:
            self.allies[side_id] = []
        if ally_id not in self.allies[side_id]:
            self.allies[side_id].append(ally_id)
        self.remove_hostile(side_id, ally_id)

    def remove_ally(self, side_id: str, ally_id: str):
        if side_id in self.allies:
            self.allies[side_id] = [id for id in self.allies[side_id] if id != ally_id]

    def is_ally(self, side_id: str, ally_id: str) -> bool:
        return ally_id in self.allies.get(side_id, [])

    def is_hostile(self, side_id: str, hostile_id: str) -> bool:
        return hostile_id in self.hostiles.get(side_id, [])

    def get_allies(self, side_id: str) -> List[str]:
        return self.allies.get(side_id, [])

    def get_hostiles(self, side_id: str) -> List[str]:
        return self.hostiles.get(side_id, [])

    def update_relationship(self, side_id: str, hostiles: List[str], allies: List[str]):
        self.hostiles[side_id] = hostiles
        self.allies[side_id] = allies

    def delete_side(self, side_id: str):
        for key in self.hostiles:
            self.hostiles[key] = [id for id in self.hostiles[key] if id != side_id]
        for key in self.allies:
            self.allies[key] = [id for id in self.allies[key] if id != side_id]
        self.hostiles.pop(side_id, None)
        self.allies.pop(side_id, None)

    def to_dict(self):
        return {
            "hostiles": self.hostiles,
            "allies": self.allies,
        }

File: client/src/game/Relationships.ts
interface IRelationships {
  hostiles?: { [sideId: string]: string[] };
  allies?: { [sideId: string]: string[] };
}

export default class Relationships {
  hostiles: { [sideId: string]: string[] };
  allies: { [sideId: string]: string[] };

  constructor(parameters: IRelationships) {
    this.hostiles = parameters.hostiles ?? {};
    this.allies = parameters.allies ?? {};
  }

  addHostile(sideId: string, hostileId: string) {
    if (!this.hostiles[sideId]) {
      this.hostiles[sideId] = [];
    }
    if (!this.hostiles[sideId].includes(hostileId)) {
      this.hostiles[sideId].push(hostileId);
    }
    this.removeAlly(sideId, hostileId);
  }

  removeHostile(sideId: string, hostileId: string) {
    if (this.hostiles[sideId]) {
      this.hostiles[sideId] = this.hostiles[sideId].filter(
        (id) => id !== hostileId
      );
    }
  }

  addAlly(sideId: string, allyId: string) {
    if (!this.allies[sideId]) {
      this.allies[sideId] = [];
    }
    if (!this.allies[sideId].includes(allyId)) {
      this.allies[sideId].push(allyId);
    }
    this.removeHostile(sideId, allyId);
  }

  removeAlly(sideId: string, allyId: string) {
    if (this.allies[sideId]) {
      this.allies[sideId] = this.allies[sideId].filter((id) => id !== allyId);
    }
  }

  isAlly(sideId: string, allyId: string): boolean {
    return this.allies[sideId]?.includes(allyId) ?? false;
  }

  isHostile(sideId: string, hostileId: string): boolean {
    return this.hostiles[sideId]?.includes(hostileId) ?? false;
  }

  getAllies(sideId: string): string[] {
    return this.allies[sideId] ?? [];
  }

  getHostiles(sideId: string): string[] {
    return this.hostiles[sideId] ?? [];
  }

  updateRelationship(sideId: string, hostiles: string[], allies: string[]) {
    this.hostiles[sideId] = hostiles;
    this.allies[sideId] = allies;
  }

  deleteSide(sideId: string) {
    Object.keys(this.hostiles).forEach((key) => {
      this.hostiles[key] = this.hostiles[key].filter((id) => id !== sideId);
    });
    Object.keys(this.allies).forEach((key) => {
      this.allies[key] = this.allies[key].filter((id) => id !== sideId);
    });
    delete this.hostiles[sideId];
    delete this.allies[sideId];
  }
}
